[Doc][KubeRay] Add a doc for RayService high-availability (#40672)
ray-project/ray: doc/source/cluster/kubernetes/user-guides/rayservice-high-availability.md

(kuberay-rayservice-ha)=

> # RayService high-availability

Reviewer: ```suggestion
# RayService high availability
```

> [RayService](kuberay-rayservice) provides high availability so that the service can still serve requests when the Ray head Pod fails.

Reviewer 2: ```suggestion
[RayService](kuberay-rayservice) provides high availability (HA) for services to continue serving requests when the Ray head Pod fails.
```

## Prerequisites

* KubeRay 1.0.0 or later
* Enable GCS fault tolerance in the RayService.

## Quickstart

### Step 1: Create a Kubernetes cluster with Kind

```sh
kind create cluster --image=kindest/node:v1.23.0
```

### Step 2: Install the KubeRay operator

Follow [this document](kuberay-operator-deploy) to install the latest stable KubeRay operator by Helm repository.

### Step 3: Install a RayService with GCS fault tolerance enabled

```sh
curl -LO https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/ray-service.high-availability.yaml
kubectl apply -f ray-service.high-availability.yaml
```

> The [ray-service.high-availability.yaml](https://raw.githubusercontent.com/ray-project/kuberay/v1.0.0/ray-operator/config/samples/ray-service.high-availability.yaml) file has several Kubernetes objects:

Author: Good point. I decide to point to the master branch before I release v1.0.0.
https://github.com/ray-project/ray/pull/40672/commits/d456ee2c3f1acc21bb6fecc7be0a65b798933452

```sh
# Port forward the Ray dashboard
kubectl port-forward --address 0.0.0.0 svc/rayservice-ha-head-svc 8265:8265
# Visit ${YOUR_IP}:8265 in your browser for the Dashboard (e.g. 127.0.0.1:8265)
# Check:
# (1) Both head and worker nodes have HTTPProxyActors.
# (2) Only worker nodes have Ray Serve replicas.
# (3) Each worker node has one Ray Serve replica for each Ray Serve deployment.
```

### Step 6: Send requests to the RayService

```sh
# Log into the separate Ray Pod.
kubectl exec -it ray-pod -- bash

# Send requests to the RayService.
python3 samples/query.py

# This script sends the same request to the RayService consecutively, ensuring at most one in-flight request at a time.
# The request is equivalent to `curl -X POST -H 'Content-Type: application/json' localhost:8000/fruit/ -d '["PEAR", 12]'`.

# [Example output]
# req_index : 2197, num_fail: 0
# response: 12
# req_index : 2198, num_fail: 0
# response: 12
# req_index : 2199, num_fail: 0
```

### Step 7: Delete the Ray head Pod

```sh
# Step 7.1: Delete the Ray head Pod.
export HEAD_POD=$(kubectl get pods --selector=ray.io/node-type=head -o custom-columns=POD:metadata.name --no-headers)
kubectl delete pod $HEAD_POD
```

> In this example, `query.py` ensures there's at most one in-flight request at any given time.

Reviewer 2: ```suggestion
In this example, `query.py` ensures that at most one request is in-flight at any given time.
```

> Furthermore, there are no Ray Serve replicas on the Ray head Pod.

Reviewer 2: ```suggestion
Furthermore, the Ray head Pod has doesn't have any Ray Serve replicas.
```

> The request may only fail when a request is in the HTTPProxyActor on the Ray head Pod.

Reviewer 2: ```suggestion
Requests may fail only when a request is in the HTTPProxyActor on the Ray head Pod.
```

> Therefore, it's highly probable that no failures occur during the deletion and recovery of the Ray head Pod.

Reviewer 2: ```suggestion
Therefore, failures are highly unlikely to occur during the deletion and recovery of the Ray head Pod.
```

> Users can implement their own retry logic in their Ray scripts to handle the failures.

Reviewer 2: ```suggestion
You can implement retry logic in Ray scripts to handle the failures.
```
